src/Std/Sat/AIG/Basic.lean

```diff
@@ -0,0 +1,506 @@
+  { input with lhs := input.lhs.cast h, rhs := input.rhs.cast h }
+
+/--
+A collection of 3 of `Ref`s, useful for `LawfulOperator`s that act on three `Ref`s at a time,
+in particular multiplexer style functions.
+-/
+structure TernaryInput (aig : AIG α) where
+  discr : Ref aig
+  lhs : Ref aig
+  rhs : Ref aig
+
+/--
+The `Ref.cast` equivalent for `TernaryInput`.
+-/
+@[inline]
+def TernaryInput.cast {aig1 aig2 : AIG α} (input : TernaryInput aig1)
+    (h : aig1.decls.size ≤ aig2.decls.size) :
+    TernaryInput aig2 :=
+  { input with discr := input.discr.cast h, lhs := input.lhs.cast h, rhs := input.rhs.cast h }
+
+/--
+An entrypoint into an `AIG`. This can be used to evaluate a circuit, starting at a certain node,
+with `AIG.denote` or to construct bigger circuits on top of this specific node.
+-/
+structure Entrypoint (α : Type) [DecidableEq α] [Hashable α] where
+  /--
+  The AIG that we are in.
+  -/
+  aig : AIG α
+  /--
+  The reference to the node in `aig` that this `Entrypoint` targets.
+  -/
+  ref : Ref aig
+
+/--
+Transform an `Entrypoint` into a graphviz string. Useful for debugging purposes.
+-/
+def toGraphviz {α : Type} [DecidableEq α] [ToString α] [Hashable α] (entry : Entrypoint α) :
+    String :=
+  let ⟨⟨decls, _, hinv⟩, ⟨idx, h⟩⟩ := entry
+  let (dag, s) := go "" decls hinv idx h |>.run .empty
+  let nodes := s.fold (fun x y ↦ x ++ toGraphvizString decls y) ""
+  "Digraph AIG {" ++ nodes ++ dag ++ "}"
+where
+  go {α : Type} [DecidableEq α] [ToString α] [Hashable α] (acc : String) (decls : Array (Decl α))
+      (hinv : IsDag α decls) (idx : Nat) (hidx : idx < decls.size)
+        : StateM (HashSet (Fin decls.size)) String := do
+    let fidx : Fin decls.size := Fin.mk idx hidx
+    if (← get).contains fidx then
+      return acc
+    modify (fun s ↦ s.insert fidx)
+    match elem : decls[idx] with
+    | Decl.const _ => return acc
+    | Decl.atom _ => return acc
+    | Decl.gate lidx ridx linv rinv =>
+      let curr := s!"{idx} -> {lidx}{invEdgeStyle linv}; {idx} -> {ridx}{invEdgeStyle rinv};"
+      let hlr := hinv hidx elem
+      let laig ← go (acc ++ curr) decls hinv lidx (by omega)
+      go laig decls hinv ridx (by omega)
+  invEdgeStyle (isInv : Bool) : String :=
+    if isInv then " [color=red]" else " [color=blue]"
+  toGraphvizString {α : Type} [DecidableEq α] [ToString α] [Hashable α] (decls : Array (Decl α))
+      (idx : Fin decls.size) : String :=
+    match decls[idx] with
+    | Decl.const b => s!"{idx} [label=\"{b}\", shape=box];"
+    | Decl.atom i => s!"{idx} [label=\"{i}\", shape=doublecircle];"
+    | Decl.gate _ _ _ _ => s!"{idx} [label=\"{idx} ∧\",shape=trapezium];"
+
+/--
+A stream of references into `aig`. This is the `AIG` analog of `BitVec`.
+-/
+structure RefStream (aig : AIG α) (w : Nat) where
+  refs : Array Nat
+  hlen : refs.size = w
+  hrefs : ∀ (h : i < w), refs[i] < aig.decls.size
+
+/--
+A sequence of references bundled with their AIG. In some sense a generalization of `Entrypoint`.
```

Comment:
```suggestion
A sequence of references bundled with their AIG. In some sense a generalization of `Entrypoint`.
```
Just drop "in some sense", or explain the sense. Otherwise it just costs reader time to work it out.